File: 0131-palindrome-partitioning/0131-palindrome-partitioning.rb
# @param {String} s
# @return {String[][]}
def pali?(st, l, r)
  while l < r
    return false if st[l] != st[r]
    l += 1
    r -= 1
  end
  true
end

def dfs(i, s, ans, part)
  if i >= s.length
    ans << part.clone
    return
  end
  (i...s.length).each do |j|
    if pali?(s, i, j)
      part << s[i..j]
      dfs(j+1, s, ans, part)
      part.pop
    end
  end
  nil
end

def partition(s)
  ans, part = [], []
  dfs(0, s, ans, part)
  ans
end

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.ts
function partition(s: string): string[][] {
  const ans: string[][] = []
  const part: string[] = []
  
  function isPali(st: string, l: number, r: number): boolean {
    while (l < r) {
      if (st[l] !== st[r]) return false
      l++
      r--
    }
    return true
  }
  
  function dfs(i: number): void {
    if (i >= s.length) {
      ans.push(part.slice())
      return
    }
    for (let j = i; j < s.length; j++) {
      if (isPali(s, i, j)) {
        part.push(s.slice(i, j + 1))
        dfs(j + 1)
        part.pop()
      }
    }
  }

  dfs(0)
  return ans
}